contract/lst-oracle/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+        assert_one_yocto();
+        self.assert_owner();
+
+        if self.transformers.insert(&price_id, &entry).is_some() {
+            env::panic_str("Price identifier collision");
+        }
+    }
+
+    // impl Pyth:
+
+    pub fn price_feed_exists(&self, price_identifier: PriceIdentifier) -> bool {
+        self.transformers.get(&price_identifier).is_some()
+    }
+
+    pub fn list_ema_prices_no_older_than(
+        &self,
+        price_ids: Vec<PriceIdentifier>,
+        age: u64,
+    ) -> Promise {
+        let (dispatched_price_ids, promises): (Vec<_>, Vec<_>) = price_ids
+            .iter()
+            .copied()
+            .map(|price_id| {
+                if let Some(entry) = self.transformers.get(&price_id) {
+                    (entry.price_id, Some(entry.call.promise()))
+                } else {
+                    (price_id, None)
+                }
+            })
+            .unzip();
+
+        let mut promise = ext_pyth::ext(self.oracle_id.clone())
+            .list_ema_prices_no_older_than(dispatched_price_ids, age);
+
+        for p in promises.into_iter().flatten() {
+            promise = promise.and(p);
+        }
+
+        promise.then(
+            self_ext!(Gas::from_tgas(4))
+                .list_ema_prices_no_older_than_01_consume_results(price_ids),
+        )
+    }
+
+    #[private]
+    pub fn list_ema_prices_no_older_than_01_consume_results(
+        &self,
+        original_price_ids: Vec<PriceIdentifier>,
+    ) -> OracleResponse {
+        fn callback_result<T: DeserializeOwned>(index: u64) -> T {
+            match env::promise_result(index) {
+                PromiseResult::Successful(vec) => serde_json::from_slice(&vec).unwrap(),
+                PromiseResult::Failed => env::panic_str(&format!("Promise index {index} failed")),
+            }
+        }
+
+        let oracle_result = callback_result::<OracleResponse>(0);
+        near_sdk::log!(
+            "Original oracle result: {}",
+            serde_json::to_string(&oracle_result).unwrap(),
+        );
+        let mut result = OracleResponse::with_capacity(oracle_result.len());
+
+        let mut i = 1;
+        for price_id in original_price_ids {
+            if let Some(price) = oracle_result.get(&price_id) {
+                near_sdk::log!("Original price passthrough: {price_id}");
+                result.insert(price_id, price.clone());
+            } else {
+                near_sdk::log!("Transforming price: {price_id}");
+                let Some(entry) = self.transformers.get(&price_id) else {
+                    env::panic_str(&format!(
+                        "No transformer associated with price ID: {price_id}",
+                    ));
+                };
+                let Some(price) = oracle_result.get(&entry.price_id) else {
+                    env::panic_str(&format!(
+                        "Mapped price ID is not in oracle result: {price_id}",
+                    ));
```

Comment:
Just a note on the logs and panic strings, in my experience strings take up a significant amount of data in a compiled contract to the point that we had a very large reduction in contract size. When we uniformed this and truncated to only be IDs, this greatly reduced the footprint.

Reply:
Yeah its that. When can get to it when we get to it, thats fine. We used stuff like "ERR_<someissue>" which was fairly general. Its possible to tack in what function it happened in as well by getting the name of the method which already is written as a string in the compiled code too for more clarity.